tsl/src/continuous_aggs/create.c

```diff
@@ -1097,38 +1097,61 @@ cagg_validate_query(Query *query, bool finalized)
 	/* check if we have a hypertable in the FROM clause */
 	rtref = linitial_node(RangeTblRef, query->jointree->fromlist);
 	rte = list_nth(query->rtable, rtref->rtindex - 1);
+
 	/* FROM only <tablename> sets rte->inh to false */
-	if (rte->relkind != RELKIND_RELATION || rte->tablesample || rte->inh == false)
+	if ((rte->relkind != RELKIND_RELATION && rte->relkind != RELKIND_VIEW) || rte->tablesample ||
+		rte->inh == false)
 	{
 		ereport(ERROR,
 				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
 				 errmsg("invalid continuous aggregate view")));
 	}
-	if (rte->relkind == RELKIND_RELATION)
+
+	if (rte->relkind == RELKIND_RELATION || rte->relkind == RELKIND_VIEW)
 	{
 		const Dimension *part_dimension = NULL;
 
-		ht = ts_hypertable_cache_get_cache_and_entry(rte->relid, CACHE_FLAG_NONE, &hcache);
+		if (rte->relkind == RELKIND_RELATION)
+			ht = ts_hypertable_cache_get_cache_and_entry(rte->relid, CACHE_FLAG_NONE, &hcache);
+		else
+		{
+			const ContinuousAgg *cagg;
+
+			cagg = ts_continuous_agg_find_by_relid(rte->relid);
+
+			if (!ContinuousAggIsFinalized(cagg))
+			{
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("not allowed old format of continuous aggregate")));
+			}
+
+			hcache = ts_hypertable_cache_pin();
+			ht = ts_hypertable_cache_get_entry_by_id(hcache, cagg->data.mat_hypertable_id);
+		}
 
 		if (TS_HYPERTABLE_IS_INTERNAL_COMPRESSION_TABLE(ht))
 			ereport(ERROR,
 					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
 					 errmsg("hypertable is an internal compressed hypertable")));
 
-		/* there can only be one continuous aggregate per table */
-		switch (ts_continuous_agg_hypertable_status(ht->fd.id))
+		if (rte->relkind == RELKIND_RELATION)
 		{
-			case HypertableIsMaterialization:
-			case HypertableIsMaterializationAndRaw:
-				ereport(ERROR,
-						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-						 errmsg("hypertable is a continuous aggregate materialization table")));
-			case HypertableIsRawTable:
-				break;
-			case HypertableIsNotContinuousAgg:
-				break;
-			default:
-				Assert(false);
+			/* there can only be one continuous aggregate per table */
+			switch (ts_continuous_agg_hypertable_status(ht->fd.id))
```

Comment:
by the way, why do we need this check at all? don't we support all these cases with this new feature?

Reply:
we need this test to prevent users create CAGG over an existing materialization hypertable. See: https://github.com/timescale/timescaledb/blob/main/tsl/test/expected/cagg_ddl.out#L467-L476